ebpf_extensions/neteventebpfext/netevent_ebpf_ext_event.c

```diff
@@ -0,0 +1,530 @@
+            netevent_event_context->event_data_end - netevent_event_context->event_data_start + 1);
+        *data_size_out = netevent_event_context->event_data_end - netevent_event_context->event_data_start;
+    } else {
+        *data_size_out = 0;
+    }
+
+    ExFreePool(netevent_event_context);
+
+Exit:
+    EBPF_EXT_LOG_EXIT();
+}
+
+//
+// Event Hook NPI client helper functions (invoked by NetEvent as the NPI provider).
+//
+typedef struct _netevent_event_notify_context
+{
+    netevent_event_md_t netevent_event_md;
+} netevent_event_notify_context_t;
+
+void
+_ebpf_netevent_push_event(_In_ netevent_event_md_t* netevent_event)
+{
+    // Logging may delay the event processing, consider enabling only for debugging or if the calling frequency for a
+    // specific use case is low.
+    // EBPF_EXT_LOG_ENTRY();
+
+    if (netevent_event == NULL) {
+        return;
+    }
+
+    ebpf_result_t result;
+    ebpf_extension_hook_client_t* client_context = NULL;
+    netevent_event_notify_context_t netevent_event_notify_context = {0};
+    uint64_t event_size = netevent_event->event_data_end - netevent_event->event_data_start;
+    bool push_lock_acquired = false;
+
+    // Currently, the verifier does not support read-only contexts, so we need to copy the event data, rather than
+    // directly passing the existing pointers.
+    // Verifier feature proposal: https://github.com/vbpf/ebpf-verifier/issues/639
+    ExAcquirePushLockExclusive(&_ebpf_netevent_push_event_lock);
+    push_lock_acquired = true;
+    if (event_size > _event_buffer_size) {
+        // If the event buffer is too small, attempt to resize it.
+        uint8_t* new_event_buffer =
+            (uint8_t*)ExAllocatePoolUninitialized(NonPagedPoolNx, event_size, EBPF_NETEVENT_EXTENSION_POOL_TAG);
+        if (new_event_buffer == NULL) {
+            EBPF_EXT_LOG_MESSAGE(
+                EBPF_EXT_TRACELOG_LEVEL_ERROR,
+                EBPF_EXT_TRACELOG_KEYWORD_NETEVENT,
+                "Failed to resize the event buffer - event lost");
+            goto Exit;
+        }
+        if (_event_buffer) {
+            ExFreePool(_event_buffer);
+        }
+        _event_buffer = new_event_buffer;
+        _event_buffer_size = event_size;
+    }
+    memcpy(_event_buffer, netevent_event->event_data_start, event_size);
+    netevent_event_notify_context.netevent_event_md.event_data_start = _event_buffer;
+    netevent_event_notify_context.netevent_event_md.event_data_end = _event_buffer + event_size;
+
+    // For each attached client call the netevent hook.
+    client_context = ebpf_extension_hook_get_next_attached_client(_ebpf_netevent_event_hook_provider_context, NULL);
+    while (client_context != NULL) {
+        NTSTATUS status = 0;
+        if (ebpf_extension_hook_client_enter_rundown(client_context)) {
+            result = ebpf_extension_hook_invoke_program(
+                client_context, &netevent_event_notify_context.netevent_event_md, (uint32_t*)&status);
```

Comment:
What is 'status' here? Do we want to add any trace here if error is set?

Reply:
The library trascelog is missing functions to log GUID & NTSTATUS, I'm already working on adding one for the GUID